Bat Files: Change batch file to compare start and end time in entirety, not just seconds

PrimusNZ opened the pull request:
Checking the seconds for is the server has been running for > 15 seconds is incredibly unreliable, especially if you have game servers that automatically restart on a schedule.

This change records the start hour and minute and uses those too to determine if the server has been up < 15 seconds

Fail Case:
Arma server which starts at 15:00:01 on a 3 hourly restart schedule. Server auto-restarts itself at 18:00:02
Since the old code was only checking the seconds, 02-01 = 01, less than 15 seconds, server was never restarted because ogp agent assumed service had been running less than 15 seconds

PrimusNZ committed:
Start Bat Files: Change batch file to compare start and end minutes and hour

own3mall commented:
Thanks for pointing this out, but your proposed solution will not work either because the start and end times will never be the same.

 We need to compare how much time has passed between the server being started and stopped.  If it is under 15 seconds, we exit and stop trying to restart the server.  This is in case there is something wrong with the server and it continuously crashes.  Normally, if a server crashes, it could happen after hours and days of running.  In this case, we want to automatically restart it after it crashes.  However, if it crashes without 15 seconds of it starting, we can assume something is very wrong and stop.

own3mall commented:
We need something like (pseudo code since I'd need to refresh my batch syntax):

```
if \"%endtimestamp%\"-  \"%starttimestamp%\" less than 15 seconds - exit
```

PrimusNZ commented:
I did find the following script we could use as a guide to correct this
```
@echo off
setlocal

rem The format of %TIME% is HH:MM:SS,CS for example 23:59:59,99
set STARTTIME=%TIME%
rem here begins the command you want to measure
pause
rem here ends the command you want to measure

set ENDTIME=%TIME%

rem output as time
echo STARTTIME: %STARTTIME%
echo ENDTIME: %ENDTIME%

rem convert STARTTIME and ENDTIME to centiseconds
set /A STARTTIME=(1%STARTTIME:~0,2%-100)*360000 + (1%STARTTIME:~3,2%-100)*6000 + (1%STARTTIME:~6,2%-100)*100 + (1%STARTTIME:~9,2%-100)
set /A ENDTIME=(1%ENDTIME:~0,2%-100)*360000 + (1%ENDTIME:~3,2%-100)*6000 + (1%ENDTIME:~6,2%-100)*100 + (1%ENDTIME:~9,2%-100)

rem calculating the duratyion is easy
set /A DURATION=%ENDTIME%-%STARTTIME%

rem we might have measured the time inbetween days
if %ENDTIME% LSS %STARTTIME% set set /A DURATION=%STARTTIME%-%ENDTIME%

rem now break the centiseconds down to hours, minutes, seconds and the remaining centiseconds
set /A DURATIONH=%DURATION% / 360000
set /A DURATIONM=(%DURATION% - %DURATIONH%*360000) / 6000
set /A DURATIONS=(%DURATION% - %DURATIONH%*360000 - %DURATIONM%*6000) / 100
set /A DURATIONHS=(%DURATION% - %DURATIONH%*360000 - %DURATIONM%*6000 - %DURATIONS%*100)

rem some formatting
if %DURATIONH% LSS 10 set DURATIONH=0%DURATIONH%
if %DURATIONM% LSS 10 set DURATIONM=0%DURATIONM%
if %DURATIONS% LSS 10 set DURATIONS=0%DURATIONS%
if %DURATIONHS% LSS 10 set DURATIONHS=0%DURATIONHS%

rem outputing
echo STARTTIME: %STARTTIME% centiseconds
echo ENDTIME: %ENDTIME% centiseconds
echo DURATION: %DURATION% in centiseconds
echo %DURATIONH%:%DURATIONM%:%DURATIONS%,%DURATIONHS%

endlocal
goto :EOF
```

PrimusNZ committed:
Update Batch script to better get time passed since server started

PrimusNZ commented:
Push an update with parts added from above, seems to work okay. Only problem that still exists is if the server runs for > 24 hours but less than 24 hours and 15 seconds. Which IMHO is far less of an issue right now then the current implementation where it completely ignores the time and just goes off the seconds.
(Its a shame windows doesnt have a seconds from epoch var/command like unix does)

own3mall reviewed (commented):
Can you try the following (modify as necessary right after the set endtime line):

```
setlocal EnableDelayedExpansion
set "end=!endtime:%time:~8,1%=%%100)*100+1!"  &  set "start=!starttime:%time:~8,1%=%%100)*100+1!"
set /A "elap=((((10!end:%time:~2,1%=%%100)*60+1!%%100)-((((10!start:%time:~2,1%=%%100)*60+1!%%100)"
rem Convert elapsed time to HH:MM:SS:CC format:
set /A "cc=elap%%100+100,elap/=100,ss=elap%%60+100,elap/=60,mm=elap%%60+100,hh=elap/60+100"
set hour=%hh:~1%
set minute=%mm:~1%
set second=%ss:~1%
IF "%hour%" == "00" IF "%minute%" == "00" IF %second% lss 15 exit
```

own3mall commented:
Sample batch file (from which my review post came from that we need to adapt our script to use):

```
@echo off
setlocal EnableDelayedExpansion

set "startTime=%time: =0%"

notepad.exe

set "endTime=%time: =0%"



rem Get elapsed time:
set "end=!endTime:%time:~8,1%=%%100)*100+1!"  &  set "start=!startTime:%time:~8,1%=%%100)*100+1!"
set /A "elap=((((10!end:%time:~2,1%=%%100)*60+1!%%100)-((((10!start:%time:~2,1%=%%100)*60+1!%%100)"

rem Convert elapsed time to HH:MM:SS:CC format:
set /A "cc=elap%%100+100,elap/=100,ss=elap%%60+100,elap/=60,mm=elap%%60+100,hh=elap/60+100"

echo Start:    %startTime%
echo End:      %endTime%
echo Elapsed:  %hh:~1%%time:~2,1%%mm:~1%%time:~2,1%%ss:~1%%time:~8,1%%cc:~1%
set hour=%hh:~1%
set minute=%mm:~1%
set second=%ss:~1%

IF "%hour%" == "00" IF "%minute%" == "00" IF %second% lss 15 ECHO LESS THAN 15 SECONDS
pause

```

PrimusNZ committed:
Change start script again to use different method of determining server runtime

PrimusNZ commented:
Seems to work also, pushing change.

Did however find though, that none of these methods seem to work correctly if the system clock isnt in 24hr

own3mall commented:
It didn't require the below line?

```
setlocal EnableDelayedExpansion
```

Were you able to actually test and see it work (crashes within 15 seconds, so it should exit)?  

When and how would the system clock not be 24hr?  You can change it to something else?  How?

PrimusNZ commented:
Ah yes, thought there was something missing from that commit.

Yeah I tested the code in a standalone script and it works as expected.

You can change the clock in the windows settings under the regional settings.

PrimusNZ commented:
Script used for testing
```
@ECHO OFF
setlocal EnableDelayedExpansion
:TOP
set STARTTIME=%time%
pause
set ENDTIME=%time%
set "end=!ENDTIME:%time:~8,1%=%%100)*100+1!"  &  set "start=!STARTTIME:%time:~8,1%=%%100)*100+1!"
set /A "elap=((((10!end:%time:~2,1%=%%100)*60+1!%%100)-((((10!start:%time:~2,1%=%%100)*60+1!%%100)"
set /A "cc=elap%%100+100,elap/=100,ss=elap%%60+100,elap/=60,mm=elap%%60+100,hh=elap/60+100"
set hour=%hh:~1%
set minute=%mm:~1%
set second=%ss:~1%
if exist SERVER_STOPPED exit
IF "%hour%" == "00" IF "%minute%" == "00" IF %second% lss 15 exit
goto TOP
```
I also adjusted the clock to make sure it worked as expected over several hours/days

PrimusNZ committed:
Add missing 'setlocal EnableDelayedExpansion'

Needs to be before the :TOP label, because if the script loops due to normal restarting, you sometimes get errors doing the setlocal - Probably only needs to be set once

own3mall commented:
Looks good.  Thanks for working on this!
